jobs: add invalid job ID checks to storage types

andyyang890 opened the pull request:
This patch aims to prevent a class of bugs resulting from jobs
unintentionally passing zero job IDs when accessing the job storage
tables, which could have unintended consequences. One way this could
conceivably happen is as a result of not initializing a field or var
that stores a job ID.

Epic: None

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/153626)
<!-- Reviewable:end -->

dt commented:
> This is important to prevent unintentional reading/writing of the same job_info rows by different jobs.

Just to clarify: you mean if two callers both pass 0, they'll read/write the same frontier (the zeroth one) right? you're not saying that zero is somehow special / causes it to iterate all the frontiers or something unusual? 

It seems ~OK to me to say you get the frontier for the id and name you pass, so if you pass 0, or 1, or 26257, that's the one you get. But I guess zero could be considered reserved as invalid  (I think when jobs are optional we use zero to mean no job) so I guess I don't have a strong objection to the special case added here even if I don't immediately see why it is worth being concerned about this specific case of passing an ID other than the one you intended.

andyyang890 commented:
> Just to clarify: you mean if two callers both pass 0, they'll read/write the same frontier (the zeroth one) right? you're not saying that zero is somehow special / causes it to iterate all the frontiers or something unusual?

Yeah, I'm imagining bugs where two jobs (e.g. different changefeeds) might accidentally both pass zero (maybe due to an uninitialized field or var) and then end up overwriting each others' progress and having undefined behavior on restore.

andyyang890 committed:
jobs: add invalid job ID checks to storage types

This patch aims to prevent a class of bugs resulting from jobs
unintentionally passing zero job IDs when accessing the job storage
tables, which could have unintended consequences. One way this could
conceivably happen is as a result of not initializing a field or var
that stores a job ID.

Release note: None
